torch/csrc/jit/passes/graph_fuser.cpp

```diff
@@ -343,8 +328,12 @@ struct GraphFuser {
     // but this requires better handling of merging fusion groups so it is not done now
     Node *real_consumer = consumer->kind() == aten::cat ? consumer->namedInput(attr::tensors)->node() : consumer;
     return isFusable(producer->node()) &&
-      allUsersAreThisConsumerOrOccurAfterIt(real_consumer, producer) &&
-      compatibleDevices(consumer, producer);
+      compatibleDevices(consumer, producer) &&
+      // Rearrange nodes such that all uses of producer are after the consumer.
+      // Fusion will rewrite those later uses to use the version of producer
+      // generated by the fused blob. In this case, producer becomes an output
+      // of the fusion group.
+      producer->node()->moveBeforeTopologicallyValid(real_consumer);
```

Comment:
Just checking: does "moveBeforeTopologicallyValid" mean "try to move this node directly before real_consumer, return false if this isn't topologically valid, return true and move the node otherwise"?

Reply:
yep! The various postcondition guarantees are documented on the method as well

Reply:
I'm a little concerned that a function called "shouldFuse" can modify the graph by moving the node. In this case it doesn't matter because how the code is written, if shouldFuse is True then we are going to modify the graph anyways, but is there something like a "canMoveBeforeTopologicallyValid" function that checks if such a move is valid?

Reply:
I agree that something called `shouldFuse` should be const-y. But after looking at the code, I'm just going to glom up `shouldFuse` and `tryFuse` (both used in only one place) into a single `tryFuse`. That way we avoid splitting conditional logic across two functions unnecessarily.